File: cpp/complex_structures/complex_data_structures.cpp
/**
 * @file complex_data_structures.cpp
 * @brief Demonstration of complex data structures in C++
 *
 * This example shows:
 * - STL containers (vector, map, set, list, etc.)
 * - Container adaptors (stack, queue, priority_queue)
 * - Custom data structures with composition
 * - Nested containers
 * - Working with iterators
 */

#include <algorithm>
#include <deque>
#include <iostream>
#include <list>
#include <map>
#include <memory>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <vector>

/**
 * @brief A simple custom data structure
 */
class Point {
public:
  Point(int x = 0, int y = 0) : x_(x), y_(y) {}

  int x() const { return x_; }
  int y() const { return y_; }

  // For comparison in sets and maps
  bool operator<(const Point &other) const {
    if (x_ == other.x_)
      return y_ < other.y_;
    return x_ < other.x_;
  }

  friend std::ostream &operator<<(std::ostream &os, const Point &p) {
    os << "(" << p.x_ << ", " << p.y_ << ")";
    return os;
  }

private:
  int x_;
  int y_;
};

/**
 * @brief A more complex data structure with composition
 */
class Shape {
public:
  virtual ~Shape() = default;
  virtual double area() const = 0;
  virtual void print() const = 0;
};

class Circle : public Shape {
public:
  Circle(const Point &center, double radius) : center_(center), radius_(radius) {}

  double area() const override { return 3.14159 * radius_ * radius_; }

  void print() const override {
    std::cout << "Circle at " << center_ << " with radius " << radius_ << std::endl;
  }

private:
  Point center_;
  double radius_;
};

class Rectangle : public Shape {
public:
  Rectangle(const Point &topLeft, const Point &bottomRight)
      : topLeft_(topLeft), bottomRight_(bottomRight) {}

  double area() const override {
    return std::abs((bottomRight_.x() - topLeft_.x()) * (bottomRight_.y() - topLeft_.y()));
  }

  void print() const override {
    std::cout << "Rectangle from " << topLeft_ << " to " << bottomRight_ << std::endl;
  }

private:
  Point topLeft_;
  Point bottomRight_;
};

/**
 * @brief Program entry point
 * @return Exit status code
 */
int main() {
  // 1. Basic vector usage
  std::cout << "1. Basic vector usage:" << std::endl;
  std::vector<int> numbers = {1, 2, 3, 4, 5};

  for (const auto &num : numbers) {
    std::cout << "  " << num << std::endl;
  }

  // 2. Map example
  std::cout << "\n2. Map example:" << std::endl;
  std::map<std::string, int> ages;
  ages["Alice"] = 30;
  ages["Bob"] = 25;
  ages["Charlie"] = 40;

  for (const auto &[name, age] : ages) {
    std::cout << "  " << name << " is " << age << " years old" << std::endl;
  }

  // 3. Set of custom objects
  std::cout << "\n3. Set of custom objects:" << std::endl;
  std::set<Point> points;
  points.insert(Point(1, 2));
  points.insert(Point(3, 4));
  points.insert(Point(1, 2)); // Duplicate, will be ignored
  points.insert(Point(5, 6));

  for (const auto &point : points) {
    std::cout << "  " << point << std::endl;
  }

  // 4. Nested containers
  std::cout << "\n4. Nested containers:" << std::endl;
  std::map<std::string, std::vector<int>> studentScores;
  studentScores["Alice"] = {90, 85, 92};
  studentScores["Bob"] = {78, 80, 85};

  for (const auto &[student, scores] : studentScores) {
    std::cout << "  " << student << "'s scores: ";
    for (const auto &score : scores) {
      std::cout << score << " ";
    }
    std::cout << std::endl;
  }

  // 5. Container adaptors
  std::cout << "\n5. Stack example:" << std::endl;
  std::stack<int> intStack;
  intStack.push(10);
  intStack.push(20);
  intStack.push(30);

  while (!intStack.empty()) {
    std::cout << "  Popped: " << intStack.top() << std::endl;
    intStack.pop();
  }

  // 6. Priority queue
  std::cout << "\n6. Priority queue example:" << std::endl;
  std::priority_queue<int> pq;
  pq.push(10);
  pq.push(30);
  pq.push(20);
  pq.push(5);

  while (!pq.empty()) {
    std::cout << "  Popped: " << pq.top() << std::endl;
    pq.pop();
  }

  // 7. Working with algorithms
  std::cout << "\n7. Algorithm example:" << std::endl;
  std::vector<int> data = {5, 2, 8, 1, 9};

  std::cout << "  Original data: ";
  for (const auto &num : data) {
    std::cout << num << " ";
  }
  std::cout << std::endl;

  std::sort(data.begin(), data.end());

  std::cout << "  Sorted data: ";
  for (const auto &num : data) {
    std::cout << num << " ";
  }
  std::cout << std::endl;

  // 8. Complex object hierarchy with smart pointers
  std::cout << "\n8. Complex object hierarchy:" << std::endl;
  std::vector<std::unique_ptr<Shape>> shapes;

  shapes.push_back(std::make_unique<Circle>(Point(0, 0), 5.0));
  shapes.push_back(std::make_unique<Rectangle>(Point(0, 0), Point(4, 3)));
  shapes.push_back(std::make_unique<Circle>(Point(5, 5), 2.5));

  for (const auto &shape : shapes) {
    shape->print();
    std::cout << "  Area: " << shape->area() << std::endl;
  }

  return 0;
}

File: python/complex_structures/complex_data_structures.py
#!/usr/bin/env python3
"""
Demonstration of complex data structures in Python

This example shows:
- Built-in container types (list, dict, set, tuple)
- Nested data structures
- Custom classes with special methods
- Collections module specialized containers
- Type annotations for complex structures
"""

from typing import List, Dict, Set, Tuple, Optional, Union, Any
from collections import defaultdict, Counter, deque, namedtuple
from dataclasses import dataclass
import json


# Named tuple - a simple way to create lightweight immutable classes
Point = namedtuple("Point", ["x", "y"])


# Dataclass - a more feature-rich way to create classes primarily for storing data
@dataclass
class Rectangle:
    """
    A rectangle defined by its top-left and bottom-right corners
    """

    top_left: Point
    bottom_right: Point

    def area(self) -> float:
        """Calculate the area of the rectangle"""
        width = self.bottom_right.x - self.top_left.x
        height = self.bottom_right.y - self.top_left.y
        return width * height

    def __str__(self) -> str:
        return f"Rectangle({self.top_left} to {self.bottom_right})"


# Regular class with special methods
class Person:
    """
    A person with a name, age, and a list of friends
    """

    def __init__(self, name: str, age: int):
        self.name = name
        self.age = age
        self.friends: List["Person"] = []

    def add_friend(self, friend: "Person") -> None:
        """Add a friend to this person's friend list"""
        self.friends.append(friend)

    def __repr__(self) -> str:
        return f"Person({self.name!r}, {self.age})"

    def __str__(self) -> str:
        return f"{self.name} ({self.age} years old)"


def demonstrate_basic_containers():
    """Demonstrate Python's basic container types"""
    print("1. Basic containers:")

    # List - ordered, mutable collection
    numbers = [1, 2, 3, 4, 5]
    print(f"  List: {numbers}")
    numbers.append(6)
    print(f"  After append: {numbers}")

    # Dictionary - key-value pairs
    person = {"name": "Alice", "age": 30, "city": "New York"}
    print(f"  Dictionary: {person}")
    person["job"] = "Engineer"
    print(f"  After adding key: {person}")

    # Set - unordered collection of unique elements
    unique_numbers = {1, 2, 3, 3, 2, 1, 4, 5}
    print(f"  Set (removes duplicates): {unique_numbers}")

    # Tuple - immutable ordered collection
    coordinates = (10, 20, 30)
    print(f"  Tuple: {coordinates}")

    # Converting between types
    print(f"  List to set: {set(numbers)}")
    print(f"  Dictionary keys to list: {list(person.keys())}")
    print(f"  Dictionary values to tuple: {tuple(person.values())}")


def demonstrate_nested_structures():
    """Demonstrate nested data structures"""
    print("\n2. Nested data structures:")

    # List of dictionaries
    students = [
        {"name": "Alice", "grades": [90, 85, 88]},
        {"name": "Bob", "grades": [75, 80, 85]},
        {"name": "Charlie", "grades": [95, 90, 92]},
    ]
    print(f"  List of dictionaries: {students}")

    # Dictionary of lists
    class_grades = {
        "Math": [90, 85, 88, 75, 80],
        "Science": [95, 85, 92, 80, 75],
        "History": [85, 90, 85, 88, 92],
    }
    print(f"  Dictionary of lists: {class_grades}")

    # Dictionary with nested dictionaries
    school = {
        "name": "Lincoln High",
        "location": {"city": "New York", "state": "NY"},
        "students": 500,
        "departments": {
            "Math": {"teachers": 5, "students": 150},
            "Science": {"teachers": 4, "students": 120},
            "History": {"teachers": 3, "students": 100},
        },
    }
    print(
        f"  Accessing nested data: {school['departments']['Math']['teachers']} Math teachers"
    )

    # Complex nesting
    university = {
        "name": "State University",
        "colleges": [
            {
                "name": "Engineering",
                "departments": [
                    {
                        "name": "Computer Science",
                        "students": [(1001, "Alice"), (1002, "Bob")],
                    },
                    {
                        "name": "Electrical",
                        "students": [(1003, "Charlie"), (1004, "Diana")],
                    },
                ],
            },
            {
                "name": "Arts",
                "departments": [
                    {"name": "History", "students": [(2001, "Eve"), (2002, "Frank")]}
                ],
            },
        ],
    }
    print(
        f"  Deeply nested access: {university['colleges'][0]['departments'][0]['students'][1][1]}"
    )


def demonstrate_specialized_containers():
    """Demonstrate specialized containers from the collections module"""
    print("\n3. Specialized containers:")

    # defaultdict - provides default values for missing keys
    word_counts = defaultdict(int)
    for word in "the quick brown fox jumps over the lazy dog".split():
        word_counts[word] += 1
    print(f"  defaultdict: {dict(word_counts)}")

    # Counter - specialized for counting items
    colors = ["red", "blue", "red", "green", "blue", "blue"]
    color_counter = Counter(colors)
    print(f"  Counter: {color_counter}")
    print(f"  Most common: {color_counter.most_common(2)}")

    # deque - double-ended queue
    queue = deque([1, 2, 3])
    queue.append(4)  # Add to right
    queue.appendleft(0)  # Add to left
    print(f"  deque: {queue}")

    # namedtuple - tuple with named fields
    p = Point(10, 20)
    print(f"  namedtuple: {p}")
    print(f"  Accessing by name: p.x = {p.x}, p.y = {p.y}")
    print(f"  Accessing by index: p[0] = {p[0]}, p[1] = {p[1]}")


def demonstrate_custom_classes():
    """Demonstrate custom classes and their usage in data structures"""
    print("\n4. Custom classes in data structures:")

    # Creating instances of our custom classes
    rect = Rectangle(Point(0, 0), Point(5, 10))
    print(f"  Rectangle: {rect}")
    print(f"  Area: {rect.area()}")

    # Using our Person class
    alice = Person("Alice", 30)
    bob = Person("Bob", 25)
    charlie = Person("Charlie", 40)

    # Creating a friend network
    alice.add_friend(bob)
    alice.add_friend(charlie)
    bob.add_friend(charlie)

    # Storing custom objects in containers
    people = [alice, bob, charlie]
    print(f"  List of custom objects: {people}")

    # Dictionary with custom objects as values
    people_dict = {person.name: person for person in people}
    print(f"  Dictionary with custom objects: {list(people_dict.keys())}")

    # Nested structure with custom objects
    friendship_graph = {
        alice: alice.friends,
        bob: bob.friends,
        charlie: charlie.friends,
    }
    print(f"  Alice's friends: {[str(friend) for friend in friendship_graph[alice]]}")


def demonstrate_type_annotations():
    """Demonstrate type annotations for complex structures"""
    print("\n5. Type annotations for complex structures:")

    # Define some complex type annotations
    print("  # Complex type definitions:")
    print("  Vector = List[float]")
    print("  Matrix = List[Vector]")
    print("  Student = Dict[str, Union[str, int, List[int]]]")
    print("  SchoolData = Dict[str, Dict[str, Union[int, List[Student]]]]")

    # Implementing the annotated functions
    print("  # Example function with complex type annotations:")
    print("  def process_student_data(students: List[Student]) -> Dict[str, float]:")
    print("      # Calculate average grades for each student")
    print(
        "      return {student['name']: sum(student['grades'])/len(student['grades'])"
    )
    print("              for student in students}")


def main():
    """Main function to run all demonstrations"""
    print("Complex Data Structures in Python\n" + "=" * 30)

    demonstrate_basic_containers()
    demonstrate_nested_structures()
    demonstrate_specialized_containers()
    demonstrate_custom_classes()
    demonstrate_type_annotations()

    print("\nAll examples completed!")


if __name__ == "__main__":
    main()
